### ISSUE TITLE
slow rand

### ISSUE THREAD

**Issue by johnnychen94** [JuliaGraphics/ColorTypes.jl#122]
I'm not sure if the slowing down is expected.

## Gray ~2x slower

```julia
julia> @benchmark rand(Gray, 255, 255)
BenchmarkTools.Trial: 
  memory estimate:  508.22 KiB
  allocs estimate:  4
  --------------
  minimum time:     164.654 μs (0.00% GC)
  median time:      327.663 μs (0.00% GC)
  mean time:        307.940 μs (8.80% GC)
  maximum time:     43.820 ms (99.16% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> @benchmark rand(Float64, 255, 255)
BenchmarkTools.Trial: 
  memory estimate:  508.14 KiB
  allocs estimate:  2
  --------------
  minimum time:     58.315 μs (0.00% GC)
  median time:      210.960 μs (0.00% GC)
  mean time:        191.754 μs (14.36% GC)
  maximum time:     42.694 ms (99.44% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> @benchmark colorview(Gray,rand(Float64, 255, 255))
BenchmarkTools.Trial: 
  memory estimate:  508.17 KiB
  allocs estimate:  3
  --------------
  minimum time:     55.410 μs (0.00% GC)
  median time:      208.160 μs (0.00% GC)
  mean time:        191.092 μs (14.42% GC)
  maximum time:     41.782 ms (99.39% GC)
  --------------
  samples:          10000
  evals/sample:     1
```

## RGB ~10x slower

```julia
julia> @benchmark rand(RGB, 255, 255)
BenchmarkTools.Trial: 
  memory estimate:  1.49 MiB
  allocs estimate:  4
  --------------
  minimum time:     2.246 ms (0.00% GC)
  median time:      2.368 ms (0.00% GC)
  mean time:        2.394 ms (1.72% GC)
  maximum time:     4.258 ms (0.00% GC)
  --------------
  samples:          2087
  evals/sample:     1

julia> @benchmark colorview(RGB, rand(Float64, 3, 255, 255))
BenchmarkTools.Trial: 
  memory estimate:  1.49 MiB
  allocs estimate:  4
  --------------
  minimum time:     164.298 μs (0.00% GC)
  median time:      189.998 μs (0.00% GC)
  mean time:        220.243 μs (12.68% GC)
  maximum time:     1.170 ms (49.08% GC)
  --------------
  samples:          10000
  evals/sample:     1
```

**Comment by timholy** [comment]
I think this is https://github.com/JuliaLang/julia/issues/28980. Possibly one could work around this with `unsafe_wrap`.

**Comment by stillyslalom** [comment]
This seems to have gotten a few orders of magnitude worse since it was opened, and the deficit is wider for fixed-point numbers:

```
julia> @btime rand(Float64, 10^3);
  1.210 μs (1 allocation: 7.94 KiB)

julia> @btime rand(Gray, 10^3);
  260.600 μs (1 allocation: 7.94 KiB)

julia> @btime rand(UInt8, 10^3);
  465.311 ns (1 allocation: 1.06 KiB)

julia> @btime rand(Gray{N0f8}, 10^3);
  261.400 μs (1 allocation: 1.06 KiB)
```

**Comment by kimikage** [comment]
Just to clarify, this is a result on Julia v1.6.0-DEV, right?

I am not inclined to maintain the "current" `rand`. However, there has been no progress in discussing the new design.

**Comment by timholy** [comment]
With https://github.com/JuliaLang/julia/pull/37230 I get

```julia
julia> @btime rand(Gray{N0f8}, 10^3);
  2.473 μs (1 allocation: 1.06 KiB)
```

**Comment by timholy** [comment]
It does seem to be a regression of 1.6, although 1.5 seems slow than with that PR:
```julia
julia> @btime rand(Gray{N0f8}, 10^3);
  3.478 μs (1 allocation: 1.06 KiB)
```

**Comment by timholy** [comment]
This has been very satisfyingly fixed on Julia 1.6.
